### ISSUE TITLE
Support Rails 5.1+ `form_with` style (`form_for` will eventually go away in Rails 6.x)

### ISSUE THREAD

**Issue by krtschmr** [bootstrap-ruby/bootstrap_form#326]
https://github.com/rails/rails/issues/25197

**Comment by mattbrictson** [comment]
Thanks for logging this issue! It sounds like this is an important change.

As for the impact to this project: Do you expect it is just a matter of adding `bootstrap_form_with` that follows the `form_with` semantics, or is there more to it than that?

**Comment by krtschmr** [comment]
it is a long road for rails. 
starting **5.1** we will have 3 methods `form_with`, `form_for`, `form_tag`. _for and _tag are soft-deprecated and propably with **6.0** or **6.1** rails will issue a deprecation-warning and then remove the methods maybe with **6.2** or later. 

actually we should implement the complete `form_with` as a `bootstrap_form_with`. i checked the form-builder and if i understand it correctly, we can easy re-use all the component-builders

**Comment by mattbrictson** [comment]
@krtschmr do you have a Rails 5.1 project that you are actively developing? And further would you use `bootstrap_form_with` on that project if it was available to you?

The reason I ask is that I have no plans in the short term to use `form_with` in my Rails projects, so I don't feel comfortable building this feature by myself. It doesn't make sense to add a feature if we have no idea whether it works in the real world.

If you have a good use case for it and would be willing to test it for me, then perhaps I will give it a shot.

**Comment by lcreid** [comment]
I have a Rails 5.1 project and I'd like to use `bootstrap_form_with` in it. The trick is, I also need to use Bootstrap 4. I notice that the `bootstrap-4` branch isn't up-to-date with some changes for Rails 5.1. Should I merge or rebase the `bootstrap-4` branch from master, and then add `bootstrap_form_with`? If so, which (merge or rebase)?

**Comment by lcreid** [comment]
I'm close to having a PR ready for this. `form_with` doesn't generate ids and classes for form elements the way the `form_for` and `form_tag` helpers did. I created a bunch of tests for `form_with`, basically testing what the existing tests for `form_for` and `form_tag` do, and they're all green, but I need to dig deeper on a couple of the cases before I feel comfortable that what I did is in fact correct. I just wanted to go on the record that I'm working on this issue, and I'm making some progress.

**Comment by lcreid** [comment]
In DHH's original Rails issue that proposed `form_with`, https://github.com/rails/rails/issues/25197, his first goal was to get rid of the default generation of DOM id and classes from the form helpers. Once `form_with` was released, some people raised an issue with the lack of default DOM ids and classes (e.g. https://github.com/rails/rails/issues/29014). However, it looks like the default behaviour of `form_with` is going to remain "no default DOM ids or classes."

The lack of default DOM id means that the way the `bootstrap_form` helpers generate labels for fields doesn't give the best user experience, since the labels are no longer clickable. I tried a few implementations to deal with it. None of them were particularly difficult. However, I think from the documentation and learning point of view, it would be better for `bootstrap_form_with` to behave like `form_with`. In this case, that means helpers under `bootstrap_form_with` should not generate any DOM ids or classes that the underlying Rails helpers wouldn't generate. 

The upshot is that users of `bootstrap_form_with` will have to specify an id for form elements for which they want the label to be clickable. This is the same change in behaviour that all Rails programmers will have to learn when switching from `form_for` to `form_with`.

There is an open PR on Rails (https://github.com/rails/rails/pull/29439) to add a `skip_default_ids = false` option to `form_with`. It would cause helpers used with `form_with` to generate default DOM ids and classes, the way they do for `form_for`. If Rails PR 29439 is merged into Rails, I believe it would work with `bootstrap_form_with`, meaning that the Rails helpers would generate default DOM ids and classes, and `bootstrap_form_with` helpers would generate labels with the expected `for` attribute. Of course, we'll have to test it if/when the PR is merged into Rails.

I hope I'll be able to submit a PR today for this issue, based on the approach I mentioned above. If anyone has any comments before I submit the PR, I'd love to hear them.
